### DISCUSSION TITLE
Acquiring data from a dataset for a MOBO process

### DISCUSSION THREAD

**Discussion by DavidSiretMarques** [meta-pytorch/botorch#1917] [Q&A]
I'm doing a MOBO development, and we have a dataset with a bunch of samples.

I'd like for the BO loop to propose samples from the dataset before starting to propose new samples that are not in the dataset (the creation of new samples is really expensive, it requires a bunch of physics experiments to be done).

I've already done something similar for a single objective BO with [MaxPosteriorSampling](https://botorch.org/api/generation.html#botorch.generation.sampling.MaxPosteriorSampling), but that doesn't work for 2 objectives. Is there any function that is able to draw samples from a dataset with 2 objectives to maximize? Or any way to adapt the functions already in BoTorch for that purpose?

Edit: the error I get when I use MaxPosteriorSampling is:
```
Exception has occurred: RuntimeError
The q-batch shape of the objective values does not agree with the q-batch shape of X. Got 2 and 360. This may happen if you used a one-to-many input transform but forgot to use a corresponding objective.
  File "D:\InSilicoOP-FUAM\In Silico OP\src\OP_pytorch_MOBO_mine.py", line 187, in optimize_qehvi
    sampled_x = sampler(to.from_numpy(full_x.values).double(), num_samples=BATCH_SIZE)
  File "D:\InSilicoOP-FUAM\In Silico OP\src\OP_pytorch_MOBO_mine.py", line 245, in MOBO
    acq_x, acq_val, acq_index = optimize_qehvi(gp_surro, X, Y, rf, BATCH_SIZE)
  File "D:\InSilicoOP-FUAM\In Silico OP\src\OP_pytorch_MOBO_mine.py", line 272, in <module>
    print(MOBO(X, Y, 100, 2, 5))
RuntimeError: The q-batch shape of the objective values does not agree with the q-batch shape of X. Got 2 and 360. This may happen if you used a one-to-many input transform but forgot to use a corresponding objective.
```

**Comment by sdaulton** [comment]
Some options would include:

1) Draw a sample of the function value at each point, and a) determine the Pareto frontier under the sampled function values and randomly sample a the desired number of points from the Pareto frontier or b) select the set of points (up to your desired batch size) that maximizes the hypervolume under the sampled function values in a sequential greedy manner. In both cases you may need to randomly sample some points to fill out the batch size if the Pareto frontier under the sampled function values is smaller than the batch size.

2) Define some fitness score the provides the fitness of each point and select points using MaxPosteriorSampling under that fitness score.

**Reply by DavidSiretMarques** [reply]
This answer ellicits more questions: 
1. How would you determine the Pareto frontier? (sorry if it's obvious, I'm kinda new to MOBO)
2. Same thing with calculating the hypervolume.

As for the second option, I think that might be easy, I'll try it out and reply with what I find

**Reply by sdaulton** [reply]
> How would you determine the Pareto frontier?
 You can use the [is_non_dominated](https://github.com/pytorch/botorch/blob/7eb847ab2f186691ec1810bfd16cdcbd38e24784/botorch/utils/multi_objective/pareto.py#L16) function.
The code to compute the pareto frontier under a single sample from the model over a discrete set would look like:
```
samples = model.posterior(X_discrete).sample().squeeze(0)
pareto_mask = is_non_dominated(samples)
pf = samples[pareto_mask]
```
Then to you could randomly sample `batch_size` points from the Pareto frontier.

Alternatively you could maximize the hypervolume in a sequential greedy fashion by using `DominatedPartitioning.compute_hypervolume()` ([link](https://github.com/pytorch/botorch/blob/7eb847ab2f186691ec1810bfd16cdcbd38e24784/botorch/utils/multi_objective/box_decompositions/dominated.py#L21)) to compute the hypervolumes. The way that this would work is that you would compute the hypervolume of each point individually, select the point with the maximum hypervolume, add that to your batch, then compute the hypervolume of the points in your batch + one additional point, and so on to build up the batch in an iterative fashion.

**Reply by Balandat** [reply]
> I'd like for the BO loop to propose samples from the dataset before starting to propose new samples that are not in the dataset (the creation of new samples is really expensive, it requires a bunch of physics experiments to be done).

I'm not sure I fully understand this part - if you already have the observations why re-propose them? Is this about noise in the evaluations? In this case you'd still have to do the experiment, right? Or is repeating the experiments with the previous parameters a lot easier / cheaper? 

Or are you simply trying to do some methods development offline first without running any experiments? In that case you may want to fit a surrogate model on your data set and use that as the synthetic ground truth for development purposes.

**Reply by DavidSiretMarques** [reply]
>Or are you simply trying to do some methods development offline first without running any experiments? In that case you may want to fit a surrogate model on your data set and use that as the synthetic ground truth for development purposes.

That's what I'm doing, what I wanted to check (thus the feeding of dataset data) is how the model improves for each new point we give it
